pkg/controller/persistentvolume/persistentvolume_claim_binder_controller.go

```diff
@@ -282,20 +283,40 @@ func syncClaim(volumeIndex *persistentVolumeOrderedIndex, binderClient binderCli
 			return fmt.Errorf("A volume match does not exist for persistent claim: %s", claim.Name)
 		}
 
-		// make a binding reference to the claim.
-		// triggers update of the claim in this controller, which builds claim status
+		// create a reference to the claim and assign it to the volume being bound.
+		// the volume is a pointer and assigning the reference fixes a race condition where another
+		// claim might match this volume but before the claimRef is persistent in the next case statement
+		claimRef, err := api.GetReference(claim)
+		if err != nil {
+			return fmt.Errorf("Unexpected error getting claim reference: %v\n", err)
+		}
+
+		// make a binding reference to the claim and ensure to update the local index to prevent dupe bindings
+		clone, err := conversion.NewCloner().DeepCopy(volume)
+		if err != nil {
+			return fmt.Errorf("Error cloning pv: %v", err)
+		}
+		volumeClone, ok := clone.(*api.PersistentVolume)
+		if !ok {
+			return fmt.Errorf("Unexpected pv cast error : %v\n", volumeClone)
+		}
+		volumeClone.Spec.ClaimRef = claimRef
+		if updatedVolume, err := binderClient.UpdatePersistentVolume(volumeClone); err != nil {
+			return fmt.Errorf("Unexpected error saving PersistentVolume.Status: %+v", err)
+		} else {
+			volumeIndex.Update(updatedVolume)
+		}
+
+		// the bind is persisted on the volume above and will always match the claim in a search.
+		// claim would remain Pending if the update fails, but processing this state is idempotent.
 		claim.Spec.VolumeName = volume.Name
-		// TODO: make this similar to Pod's binding both with BindingREST subresource and GuaranteedUpdate helper in etcd.go
 		claim, err = binderClient.UpdatePersistentVolumeClaim(claim)
 		if err == nil {
 			nextPhase = api.ClaimBound
 			glog.V(5).Infof("PersistentVolumeClaim[%s] is bound\n", claim.Name)
 		} else {
-			// Rollback by unsetting the ClaimRef on the volume pointer.
-			// the volume in the index will be unbound again and ready to be matched.
+			// Rollback
 			claim.Spec.VolumeName = ""
```

Comment:
This is now a no-op.  The update in this method is gated on a condition that won't be satisfied https://github.com/kubernetes/kubernetes/pull/16432/files#diff-97a1ffdd78274d54ba9dc28656ceca2dR381 .  

Why not refactor with:
```go
claim.Spec.VolumeName = volume.Name
claim.Status.Phase = api.ClaimBound
if err := binderClient.UpdatePersistentVolumeClaim(claim); err != nil{
    return err
}
```

Since you can't guarantee and update one way or the other, if that code isn't safe, this method isn't safe.